src/FunctionApps/python/IntakePipeline/linkage.py

```diff
@@ -1,2 +1,55 @@
-def add_patient_identifier(bundle: dict) -> None:
-    pass
+import hashlib
+
+
+def add_patient_identifier(salt_suffix: str, bundle: dict) -> None:
+    """
+    Given a FHIR resource bundle (defined as a dictionary of resources
+    containing at least one patient resource):
+      - identify the patient resource(s) in the bundle
+      - extract standardized name, DOB, and address information for each
+      - compute a unique hash string based on these fields
+      - add the hash string to the list of identifiers held in that
+        patient resource
+    This function assumes data has been standardized already by the
+    silver transforms.
+    """
+    for resource in bundle["entry"]:
+        if resource["resource"]["resourceType"] == "Patient":
+            patient = resource["resource"]
+
+            # Combine given and family name
+            recent_name = patient["name"][0]
+            name_str = "-".join(recent_name["given"]) + "-" + recent_name["family"]
+
+            # Compile one-line address string
+            address = patient["address"][0]
+            address_line = " ".join(address.get("line"))
+            address_line += f" {address.get('city')}, {address.get('state')}"
+            if "postalCode" in address and address["postalCode"]:
+                address_line += f" {address['postalCode']}"
+
+            # Generate and store unique hash code
+            link_str = name_str + "-" + patient["birthDate"] + "-" + address_line
+            hashcode = generate_hash_str(link_str, salt_suffix)
+            patient["identifier"].append(
+                {
+                    "value": hashcode,
+                    # Note: this system value corresponds to the FHIR specification
+                    # for a globally used / generated ID or UUID--the standard here
+                    # is to make the use "temporary" even if it's not
+                    "system": "urn:ietf:rfc:3986",
```

Comment:
It's not a URI - is there anything that says we can't just make up a system so it doesn't clash? Maybe even something that points back to us specifically?

Reply:
I know it's not a URI but that's the FHIR specification for what to put in the system field if you're rolling your own "global" or otherwise custom identifier. If we were to publish an organization page and then fill our some documented information on there we could def point it back at us, but short of that, everything I've read in the docs says to just use this so that others working with the data know we basically made this up.

Reply:
For some insight into how FHIR thinks about ID systems, check out their explanation of the system parameter under the identifier table: https://www.hl7.org/fhir/datatypes.html#Identifier

And then here are the applicable types of identifier systems they accept: https://www.hl7.org/fhir/identifier-registry.html. We can't use a URL because that has to resolve, and we don't really fit any other category. And in the context of our application, patients do theoretically have to get unique hashes generated for them. I agree it's not a perfect fit but seems like the best we can do given their specc?